src/OpenStudy/study_openinterface.jl

```diff
@@ -260,47 +435,139 @@ function get_parms(
+    ::Type{T};
+    check_type::Bool = true,
+    check_parm::Bool = true,
+    ignore::Bool = false,
+    default::T = _default_value(T),
+) where T
+
+    attr_struct = get_attribute_struct(data, col, name)
     if check_type
-        if name == "name" || name == "AVId" # TODO? protect spelling
-            if String != T
-                error("Attribute code is of type String.")
-            end
-        elseif name == "code"
-            if Int32 != T
-                error("Attribute code is of type Int32.")
-            end
-        else
-            _type = data.data_struct[col][name].type
-            if _type != T
-                error("Attribute $name of collection $col is a of type $(_type) not $T.")
-            end
-        end
+        _check_type(attr_struct, T, col, name)
     end
-    if check_parm && !(name in ["name", "code", "AVId"])
-        if data.data_struct[col][name].is_vector
-            error("Attribute $name of collection $col is a of type vector. Use `mapped_vector` instead.")
-        end
+    if check_parm
+        _check_parm(attr_struct, col, name)
     end
 
-    out = T[default for _ in 1:n]
+    n = max_elements(data, col)
+    out = Vector{T}[]
+    sizehint!(out, n)
+    for i in 1:n
+        push!(out, get_parm_1d(data, col, name, i, T; default = default))
+    end
+    return out
+end
 
-    for (idx, el) in enumerate(raw[col])
-        if !haskey(el, name)
-            error("Attribute $name not found in collection $col")
-        end
-        out[idx] = el[name] #::T # assert type
+function get_parms_2d(
+    data::Data,
+    col::String,
+    name::String,
+    ::Type{T};
+    check_type::Bool = true,
+    check_parm::Bool = true,
+    ignore::Bool = false,
+    default::T = _default_value(T),
+) where T
+
+    attr_struct = get_attribute_struct(data, col, name)
+    if check_type
+        _check_type(attr_struct, T, col, name)
+    end
+    if check_parm
+        _check_parm(attr_struct, col, name)
     end
 
+    n = max_elements(data, col)
+    out = Matrix{T}[]
+    sizehint!(out, n)
+    for i in 1:n
+        push!(out, get_parm_2d(data, col, name, i, T; default = default))
+    end
     return out
 end
 
+function get_attribute_dim1(
+    data::Data,
+    col::String,
+    attribute::String,
+    index::Integer;
+    max_check::Integer = 100,
+)
+    attr_struct = get_attribute_struct(data, col, attribute)
+    dim = attr_struct.dim
+    if dim == 0
+        error("Attribute $attribute from collection $col has no extra dimensions")
+    end
+    _check_element_range(data, col, index)
+
+    raw = _raw(data)
+
+    element = raw[col][index]
+
+    for i in 1:max_check
+        query_name = if dim == 1
+            attribute * "($i)"
+        elseif dim == 2
+            attribute * "($i,1)"
+        end
+        if !haskey(element, query_name)
+            return i - 1
+        end
+    end
+    error("Attribute $attribute from collection $col has dim1 larger than $max_check")
+    return 999
```

Comment:
This 999 is kind of weird, it never reaches this part because it errors before. We could wrap this 999 into a function `invalid_something`.

Reply:
It never reaches this.
Just added there because I never know what happens if I remove the line. Will there be type instability?